File: source/library/Spring.js
//Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt

function Spring(startpos,endpos,m,r,k,time,pause)
{
    this.osc           	=   KDampedOscillator.Create( m || 1,  r || 12,  k || 100,0,0) ;
    this.osctimer		=	Ktimer.Create(1,pause) ;
	this.time			= 	0 ;
	this.startpos		=	startpos ;
	this.endpos			=	endpos ;
	this.finishby		=	time || 5 ;
	
	this.osc.reset( this.startpos,  0,  this.endpos) ;
	
}

Spring.Create = function(startpos,endpos,m,r,k,time,pause)
{
    return new Spring(startpos,endpos,m,r,k,time,pause) ;
}

Spring.ZEROVEL_EPS		=	0.01 ;

Spring.prototype = {

	constructor: Spring,
	
	Resume: function()
	{
		this.osctimer.resume() ;
	},

   	IsFinished: function()
	{
        return (this.finishby && this.time > this.finishby) ;
    },

	IsStationary: function()
	{
		 return this.osc.equilibrium(this.osctimer.elapsed(),Spring.ZEROVEL_EPS) ;
	},

   	Update: function(dt)
	{
		this.time	=	this.time + dt
//		if (Spring.IsFinished(spring)) {
//            spring.osc.reset(spring.endpos, 0, spring.endpos)
//        }
    },

  	GetPosition: function()
	{
		if (this.IsFinished()) 
		{
			return this.endpos ;
		}
		else
		{
			return this.osc.evaluate( this.osctimer.elapsed()) ;
		}
   },
	
	toString: function()
	{
		return "Spring" ;	
	}
	


}

File: lua/Spring.lua
-- Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt
if (not Spring) then

Spring					=	{}
Spring.className	    =	"Spring"
Spring.ZEROVEL_EPS		=	0.01

    function Spring.Create(startpos,endpos,m,r,k,time,pause)

        local spring = {
            osc             =   KDampedOscillator.Create( m or 1,  r or 12,  k or 100,0,0),
            osctimer        =	Ktimer.Create(1,pause),
			time			= 	0,
			startpos		=	startpos,
			endpos			=	endpos,
			finishby		=	time or 5,
            className       =   Spring.className,
        }
        setmetatable(spring,{__index = Spring } )
		spring.osc:reset( startpos,  0,  endpos)
        return spring
    end

	function Spring.Resume(spring)
		Ktimer.resume(spring.osctimer)
	end

    function Spring.IsFinished(spring)
        return (spring.finishby and spring.time > spring.finishby)
    end

	function Spring.IsStationary(spring)
		 return KDampedOscillator.equilibrium(spring.osc,Ktimer.elapsed(spring.osctimer),Spring.ZEROVEL_EPS)
	end

    function Spring.Update(spring,dt)
		spring.time	=	spring.time + dt
--		if (Spring.IsFinished(spring)) then
--            spring.osc:reset(spring.endpos, 0, spring.endpos)
--        end
    end

    function Spring.GetPosition(spring)
		if (Spring.IsFinished(spring)) then
			return spring.endpos
		else
			return KDampedOscillator.evaluate(spring.osc, Ktimer.elapsed(spring.osctimer))
		end
    end

end
